<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>PSF: Pathway Signal Flow Calculator â€¢ psf</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="PSF: Pathway Signal Flow Calculator">
<meta name="description" content="Calculates activity of KEGG signaling pathways based on expression fold change values and interaction types.">
<meta property="og:description" content="Calculates activity of KEGG signaling pathways based on expression fold change values and interaction types.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">psf</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/Intro_to_psf.html">PSF toolkit: Getting Started</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/hakobyansiras/PSFC"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="psf-toolkit">PSF toolkit<a class="anchor" aria-label="anchor" href="#psf-toolkit"></a>
</h1></div>
<!-- badges: start -->

<p>The <strong>PSF Toolkit</strong> is an R package developed for <strong>topology-aware (TA) pathway analysis</strong> of various types of omics data. The package includes interactive modules for <strong>pathway editing</strong> and <strong>visualization</strong>, facilitating pathway curation and results interpretation.</p>
<ul>
<li>Website: <a href="https://hakobyansiras.github.io/PSFC" class="external-link uri">https://hakobyansiras.github.io/PSFC</a>
</li>
<li>Get started: see the Articles tab on the website.</li>
</ul>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"hakobyansiras/psf"</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><strong>Note:</strong> Due to potential network issues with GitHub servers, the package download may take longer than expected. If you encounter timeout errors, try increasing the timeout limit using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span></code></pre></div>
</blockquote>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level3">
<h3 id="running-the-shiny-app-for-pathway-curation-and-visualization">Running the Shiny App for Pathway Curation and Visualization<a class="anchor" aria-label="anchor" href="#running-the-shiny-app-for-pathway-curation-and-visualization"></a>
</h3>
<p>The PSF Toolkit includes a <strong>Shiny</strong> app for interactive pathway curation and visualization.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hakobyansiras/PSFC" class="external-link">psf</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/run_shiny_app.html">run_shiny_app</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>A detailed tutorial on pathway import, curation, and analysis is available on the website.</p>
</div>
</div>
<div class="section level2">
<h2 id="build-documentation-site-locally">Build documentation site locally<a class="anchor" aria-label="anchor" href="#build-documentation-site-locally"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install development dependencies</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"roxygen2"</span>, <span class="st">"pkgdown"</span><span class="op">)</span>, repos <span class="op">=</span> <span class="st">"https://cloud.r-project.org"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate Rd files (optional, already present)</span></span>
<span><span class="fu">roxygen2</span><span class="fu">::</span><span class="fu"><a href="https://roxygen2.r-lib.org/reference/roxygenize.html" class="external-link">roxygenise</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Build site into docs/</span></span>
<span><span class="fu">pkgdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgdown.r-lib.org/reference/build_site.html" class="external-link">build_site</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-psf-on-bulk-rna-seq-data-with-40-curated-kegg-signaling-pathways">
<strong>Running PSF on Bulk RNA-seq Data with 40 Curated KEGG Signaling Pathways</strong><a class="anchor" aria-label="anchor" href="#running-psf-on-bulk-rna-seq-data-with-40-curated-kegg-signaling-pathways"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hakobyansiras/PSFC" class="external-link">psf</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load melanoma DESeq2 normalized expression data (GSE112509)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"melanoma_exp.RData"</span>, package<span class="op">=</span><span class="st">"psf"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load curated KEGG pathway collection</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"kegg_curated_40_signalings.RData"</span>, package<span class="op">=</span><span class="st">"psf"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate fold change (FC) values against the global mean</span></span>
<span><span class="va">melanoma_fc</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">melanoma_deseq_normalized_counts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">melanoma_deseq_normalized_counts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run PSF analysis on 40 pathways for all samples</span></span>
<span><span class="va">melanoma_psf</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/run_psf.html">run_psf</a></span><span class="op">(</span>entrez.fc <span class="op">=</span> <span class="va">melanoma_fc</span>, kegg.collection <span class="op">=</span> <span class="va">kegg_curated_40_signalings</span>, </span>
<span>                        calculate.significance <span class="op">=</span> <span class="cn">FALSE</span>, ncores <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot KEGG image-based PI3K-Akt signaling pathway with color-coded nodes</span></span>
<span><span class="fu"><a href="reference/plot_pathway.html">plot_pathway</a></span><span class="op">(</span><span class="va">melanoma_psf</span><span class="op">$</span><span class="va">PI3K_Akt_signaling_pathway</span>, plot_type <span class="op">=</span> <span class="st">"kegg"</span>, </span>
<span>             color_nodes <span class="op">=</span> <span class="st">"psf_activities"</span>, sample_id <span class="op">=</span> <span class="st">"mean"</span>, log_norm <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>             plot_sink_values <span class="op">=</span> <span class="cn">TRUE</span>, use_old_images <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate PDF reports for PSF-analyzed pathways</span></span>
<span><span class="fu"><a href="reference/generate_psf_report.html">generate_psf_report</a></span><span class="op">(</span>psf_list <span class="op">=</span> <span class="va">melanoma_psf</span>, folder_name <span class="op">=</span> <span class="st">"example_psf_report"</span>, </span>
<span>                    plot_type <span class="op">=</span> <span class="st">"kegg"</span>, log_norm <span class="op">=</span> <span class="cn">TRUE</span>, use_old_images <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-psf-on-spatial-transcriptomics-data">
<strong>Running PSF on Spatial Transcriptomics Data</strong><a class="anchor" aria-label="anchor" href="#running-psf-on-spatial-transcriptomics-data"></a>
</h2>
<blockquote>
<p><strong>Important:</strong> Due to installation issues with the <code>Seurat</code> package, it is included in <code>Suggests</code> rather than <code>Imports</code>. The PSF package will still install even if <code>Seurat</code> fails to install. However, you must manually install <code>Seurat</code> to perform spatial transcriptomic analyses.</p>
</blockquote>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hakobyansiras/PSFC" class="external-link">psf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the spatial dataset. The human melanoma dataset can be downloaded from the following link: https://www.10xgenomics.com/datasets/human-melanoma-if-stained-ffpe-2-standard. To import the data into Seurat, please download the Feature / barcode matrix HDF5 (filtered) and the Spatial imaging data files.</span></span>
<span><span class="va">spatial_melanoma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Load10X_Spatial.html" class="external-link">Load10X_Spatial</a></span><span class="op">(</span><span class="st">"/path/to/spatial_dataset"</span>, </span>
<span>                                    filename <span class="op">=</span> <span class="st">"CytAssist_FFPE_Human_Skin_Melanoma_filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load pre-downloaded gene symbol to Entrez ID conversion data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"gene_symbol_to_entrez.RData"</span>, package<span class="op">=</span><span class="st">"psf"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load curated KEGG signaling pathways</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"kegg_curated_40_signalings.RData"</span>, package<span class="op">=</span><span class="st">"psf"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"kegg_sink_to_process.RData"</span>, package<span class="op">=</span><span class="st">"psf"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run pathway analysis</span></span>
<span><span class="va">psf_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/spatial_psf_analysis.html">spatial_psf_analysis</a></span><span class="op">(</span>spatial_obj <span class="op">=</span> <span class="va">spatial_melanoma</span>, </span>
<span>                                    pathway_collection <span class="op">=</span> <span class="va">kegg_curated_40_signalings</span>, </span>
<span>                                    gene_symbol_to_entrez <span class="op">=</span> <span class="va">gene_symbol_to_entrez</span>, nthreads <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Launch the PSF spatial browser application</span></span>
<span><span class="fu"><a href="reference/run_psf_spatial_browser.html">run_psf_spatial_browser</a></span><span class="op">(</span>psf_spatial_results <span class="op">=</span> <span class="va">psf_spatial</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="app-usage">App Usage<a class="anchor" aria-label="anchor" href="#app-usage"></a>
</h3>
<p>A short demo video demonstrating the PSF spatial browser is available: <a href="https://www.youtube.com/watch?v=lHTgYBA374o" class="external-link">PSF Spatial Browser Demo</a></p>
<p>You can also explore a <strong>demo Shiny app</strong>: <a href="https://apps.armlifebank.am/PSF_spatial_browser/" class="external-link">PSF Spatial Browser</a></p>
<p>The app includes a <strong>dedicated Help tab</strong> containing detailed instructions on how to use it.</p>
</div>
</div>
<div class="section level2">
<h2 id="running-psf-on-telomere-maintenance-mechanism-tmm-pathways">
<strong>Running PSF on Telomere Maintenance Mechanism (TMM) Pathways</strong><a class="anchor" aria-label="anchor" href="#running-psf-on-telomere-maintenance-mechanism-tmm-pathways"></a>
</h2>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hakobyansiras/PSFC" class="external-link">psf</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load melanoma DESeq2 normalized expression data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"melanoma_exp.RData"</span>, package<span class="op">=</span><span class="st">"psf"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load TMM pathway</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tmm_pathway.RData"</span>, package<span class="op">=</span><span class="st">"psf"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate fold change (FC) values against the global mean</span></span>
<span><span class="va">melanoma_fc</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">melanoma_deseq_normalized_counts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">melanoma_deseq_normalized_counts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run PSF analysis on the TMM pathway for all samples</span></span>
<span><span class="va">psf_tmm_result_mel</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/psf.from.env.entrez.fc.html">psf.from.env.entrez.fc</a></span><span class="op">(</span>entrez.fc <span class="op">=</span> <span class="va">melanoma_fc</span>,</span>
<span>                                             kegg.collection <span class="op">=</span> <span class="va">tmm_pathway</span>, </span>
<span>                                             calculate.significance <span class="op">=</span> <span class="cn">FALSE</span>, sum <span class="op">=</span> <span class="cn">FALSE</span>, split <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                             return_only_signals <span class="op">=</span> <span class="cn">FALSE</span>, tmm_mode <span class="op">=</span> <span class="cn">TRUE</span>, tmm_updated_mode <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the TMM pathway for a sample</span></span>
<span><span class="fu"><a href="reference/plot_tmm_pathway.html">plot_tmm_pathway</a></span><span class="op">(</span>pathway <span class="op">=</span> <span class="va">psf_tmm_result_mel</span><span class="op">$</span><span class="va">X031</span><span class="op">$</span><span class="va">tmm</span>, no_color_mode <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                 mapping_data_type <span class="op">=</span> <span class="st">"signal"</span>, log_norm <span class="op">=</span> <span class="cn">FALSE</span>, layout <span class="op">=</span> <span class="st">"layout_nicely"</span><span class="op">)</span></span></code></pre></div>
<p>For more details, please refer to the package documentation and tutorials.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/hakobyansiras/PSFC/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Siras Hakobyan, Lilit Nersisyan.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
